mbed-client-libservice/m2mdynmemLIB.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2014-2016 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/**
+ * \file m2mdynmemLIB.h
+ * \brief Dynamical Memory API for library model
+ *
+ */
+#ifndef M2MDYNMEMLIB_H_
+#define M2MDYNMEMLIB_H_
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "ns_types.h"
+
+/*!
+ * \enum heap_fail_t
+ * \brief Dynamically heap system failure call back event types.
+ */
+typedef enum {
+    M2M_DYN_MEM_NULL_FREE,               /**< m2m_dyn_mem_free(), NULL pointer free [obsolete - no longer faulted]  */
+    M2M_DYN_MEM_DOUBLE_FREE,                     /**< m2m_dyn_mem_free(), Possible double pointer free */
+    M2M_DYN_MEM_ALLOCATE_SIZE_NOT_VALID, /**< Allocate size is 0 or smaller or size is bigger than max heap size  */
+    M2M_DYN_MEM_POINTER_NOT_VALID,       /**< m2m_dyn_mem_free(), try to free pointer which not at heap sector */
+    M2M_DYN_MEM_HEAP_SECTOR_CORRUPTED,   /**< Heap system detect sector corruption */
+    M2M_DYN_MEM_HEAP_SECTOR_UNITIALIZED /**< m2m_dyn_mem_free(), m2m_dyn_mem_temporary_alloc() or m2m_dyn_mem_alloc() called before m2m_dyn_mem_init() */
+} m2m_heap_fail_t;
+
+typedef uint16_t m2m_block_size_t; //external interface unsigned heap block size type
+typedef uint16_t m2m_heap_size_t; //total heap size type.
+/* With current implementation m2m_block_size_t and m2m_heap_size_t should be same size */
+
+/**
+ * /struct mem_stat_t
+ * /brief Struct for Memory stats Buffer structure
+ */
+typedef struct m2m_mem_stat_t {
+    /*Heap stats*/
+    int16_t heap_sector_size;                /**< Heap total Sector len. */
+    int16_t heap_sector_alloc_cnt;                  /**< Reserved Heap sector cnt. */
+    int16_t heap_sector_allocated_bytes;     /**< Reserved Heap data in bytes. */
+    int16_t heap_sector_allocated_bytes_max; /**< Reserved Heap data in bytes max value. */
+    uint32_t heap_alloc_total_bytes;                 /**< Total Heap allocated bytes. */
+    uint32_t heap_alloc_fail_cnt;                    /**< Counter for Heap allocation fail. */
+} m2m_mem_stat_t;
+
+struct book;
```

Comment:
Now this is public API it needs a prefix matching the rest. And it's better to have it as a public typedef (while keeping the actual structure hidden).

On that note not massively fond of the m2m naming in libservice. We do need to name it differently somehow for compatibility. Only other thing I can think of is just a subtle change like dropping the pointless "dyn" - "nsmemlib.h". Not that the "ns" prefix is brilliant either, but matches the rest of the surrounding code.

But I could live with m2mdynmemlib.